include/zephyr/storage/storage_area/storage_area.h

```diff
@@ -0,0 +1,213 @@
+ * The macro definitions xxxxx_storage_area(...) checks these conditions but
+ * always succeeds. Trying to use a badly sized storage area will result in
+ * failure of any of the exposed methods.
+ *
+ */
+
+#ifndef ZEPHYR_INCLUDE_STORAGE_STORAGE_AREA_H_
+#define ZEPHYR_INCLUDE_STORAGE_STORAGE_AREA_H_
+
+#include <stdint.h>
+#include <stddef.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Storage_area interface
+ * @defgroup storage_area_interface Storage_area interface
+ * @ingroup storage_apis
+ * @{
+ */
+
+struct storage_area;
+
+struct storage_area_chunk { /* storage area data chunk */
+	void *data;         /* pointer to data */
+	size_t len;         /* data length */
+};
+
+enum storage_area_properties_mask {
+	SA_PROP_READONLY = 0x0001,
+	SA_PROP_FOVRWRITE = 0x0002, /* full overwrite (ram, rram, ...) */
+	SA_PROP_LOVRWRITE = 0x0004, /* limited overwrite (nor flash) */
+	SA_PROP_ZEROERASE = 0x0008, /* erased value is 0x00 */
+};
+
+enum storage_area_ioctl_cmd {
+	SA_IOCTL_NONE,
+	SA_IOCTL_XIPADDRESS, /* retrieve the storage area xip address */
+};
+
+/**
+ * @brief storage_area API
+ *
+ * API to access storage area.
+ */
+struct storage_area_api {
+	int (*read)(const struct storage_area *area, size_t start,
+		    const struct storage_area_chunk *ch, size_t cnt);
+	int (*prog)(const struct storage_area *area, size_t start,
+		    const struct storage_area_chunk *ch, size_t cnt);
+	int (*erase)(const struct storage_area *area, size_t start, size_t bcnt);
+	int (*ioctl)(const struct storage_area *area,
+		     enum storage_area_ioctl_cmd cmd, void *data);
+};
+
+/**
+ * @brief storage_area struct
+ */
+struct storage_area {
+	const struct storage_area_api *api;
+	size_t write_size;
+	size_t erase_size;
+	size_t erase_blocks;
+	uint32_t props; /* bitfield of storage area properties */
+};
+
+/**
+ * @brief storage_area macros
+ */
+#define STORAGE_AREA_HAS_PROPERTY(area, prop) ((area->props & prop) == prop)
+#define STORAGE_AREA_WRITESIZE(area)          area->write_size
+#define STORAGE_AREA_ERASESIZE(area)          area->erase_size
+#define STORAGE_AREA_SIZE(area)               area->erase_size * area->erase_blocks
+#define STORAGE_AREA_FOVRWRITE(area)                                            \
+	STORAGE_AREA_HAS_PROPERTY(area, SA_PROP_FOVRWRITE)
+#define STORAGE_AREA_LOVRWRITE(area)                                            \
+	STORAGE_AREA_HAS_PROPERTY(area, SA_PROP_LOVRWRITE)
+#define STORAGE_AREA_ERASEVALUE(area)                                           \
+	STORAGE_AREA_HAS_PROPERTY(area, SA_PROP_ZEROERASE) ? 0x00 : 0xff
+
+/**
+ * @brief	Read storage chunks.
+ *
+ * @param area	storage area.
+ * @param start	start in storage area (byte).
+ * @param ch	chunks for read.
+ * @param cnt   chunk element count.
+ *
+ * @retval	0 on success else negative errno code.
+ */
+int storage_area_read(const struct storage_area *area, size_t start,
+		      const struct storage_area_chunk *ch, size_t cnt);
```

Comment:
Rather than read / dread and prog / dprog, it might be better to use readv / read and writev / write as verbs, since they're more conventional.

Reply:
Thank you for this comment. I've never come accross readv/writev (never stop learning...) but you are completely right. Wonder if I should also use `iovec` instead of `storage_area_chunk` (another self invented terminology).

Reply:
Maybe - I would definitely still prefix the functions with `storage_..`.

One possibility is to use the `zvfs_` prefix for a native `struct zvfs_iovec`.